subxt/src/events/events_type.rs

```diff
@@ -206,138 +119,163 @@ impl<'a, T: Config, Evs: Decode> Events<T, Evs> {
         Ok(self.find::<Ev>().next().transpose()?.is_some())
     }
 }
 
-/// A decoded event and associated details.
+/// The event details.
 #[derive(Debug, Clone, PartialEq)]
-pub struct EventDetails<Evs> {
-    /// During which [`Phase`] was the event produced?
-    pub phase: Phase,
-    /// What index is this event in the stored events for this block.
-    pub index: u32,
-    /// The event itself.
-    pub event: Evs,
+pub struct EventDetails {
+    phase: Phase,
+    index: u32,
+    pallet: String,
+    variant: String,
+    bytes: Vec<u8>,
+    // Dev note: this is here because we've pretty much had to generate it
+    // anyway, but expect it to be generated on the fly in future versions,
+    // and so don't expose it.
+    fields: Vec<(Option<String>, DecodedValue)>,
 }
 
-/// A Value which has been decoded from some raw bytes.
-pub type DecodedValue = scale_value::Value<scale_value::scale::TypeId>;
-
-/// The raw bytes for an event with associated details about
-/// where and when it was emitted.
+/// The raw data associated with some event.
 #[derive(Debug, Clone, PartialEq)]
-pub struct RawEventDetails {
+pub struct EventDetailParts {
     /// When was the event produced?
     pub phase: Phase,
     /// What index is this event in the stored events for this block.
     pub index: u32,
     /// The name of the pallet from whence the Event originated.
     pub pallet: String,
-    /// The index of the pallet from whence the Event originated.
-    pub pallet_index: u8,
-    /// The name of the pallet Event variant.
+    /// The name of the pallet's Event variant.
     pub variant: String,
-    /// The index of the pallet Event variant.
-    pub variant_index: u8,
-    /// The bytes representing the fields contained within the event.
+    /// All of the bytes representing this event, including the pallet
+    /// and variant index that the event originated from.
     pub bytes: Vec<u8>,
-    /// Generic values representing each field of the event.
-    pub fields: Vec<DecodedValue>,
 }
 
-impl RawEventDetails {
-    /// Attempt to decode this [`RawEventDetails`] into a specific event.
-    pub fn as_event<E: Event>(&self) -> Result<Option<E>, CodecError> {
+impl EventDetails {
+    /// Return the raw data associated with this event. Useful if you want
+    /// ownership over parts of the event data.
+    pub fn parts(self) -> EventDetailParts {
+        EventDetailParts {
+            phase: self.phase,
+            index: self.index,
+            pallet: self.pallet,
+            variant: self.variant,
+            bytes: self.bytes,
+        }
+    }
+
+    /// When was the event produced?
+    pub fn phase(&self) -> Phase {
+        self.phase
+    }
+
+    /// What index is this event in the stored events for this block.
+    pub fn index(&self) -> u32 {
+        self.index
+    }
+
+    /// The index of the pallet that the event originated from.
+    pub fn pallet_index(&self) -> u8 {
```

Comment:
Actually, they should never panic (see line 302: `let mut event_bytes = vec![pallet_index, variant_index];`); we'll catch such errors when decoding into this thing. But I should add a comment or something here to make that clear, since it's a legitimate concern looking at the code for sure!